### DISCUSSION TITLE
How to add event handler to the monitor plugin/unplug events?

### DISCUSSION THREAD

**Discussion by Bo-Liu-BL** [AvaloniaUI/Avalonia#16683] [Q&A] [answered]
Hi experts,

In WPF, we can add handler to ``Microsoft.Win32.SystemEvents.DisplaySettingsChanging`` when there are monitor plugin/unplug events.
For example, we can show our window in specified monitors after the monitor changes.

**In Avalonia, is there any good way to achieve that? Assume that I can use the latest Avalonia. Thank you!**

I checked the source code, looks we can use `Win32Properties.AddWndProcHookCallback`.
The steps are:
1. Set `Win32Properties.AddWndProcHookCallback` with the main window.
```
if (Avalonia.Application.Current.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime mainWindow)
{
         Win32Properties.AddWndProcHookCallback(mainWindow.MainWindow,
                                                MyWndProcHookCallback);
}
```

2. In `MyWndProcHookCallback`, process the Windows event `WM_DISPLAYCHANGE`:
```
   public IntPtr WndProcHookCallback(IntPtr hWnd, uint msg, IntPtr wParam, IntPtr lParam, ref bool handled)
   {
      switch ((uint)msg)
      {
         case WM_DISPLAYCHANGE:
            // XXX my code logic here.
            MyHandler();
            break;
         default:
            break;
      }
      return IntPtr.Zero;
   }
```

I tested the above way, and it worked. When I add or remove monitors in my Windows system, `MyHandler` will be called.

**My question are:**
1. In Avalonia, is it fine or a good practice to use `Win32Properties.AddWndProcHookCallback` to detect and process the Windows event `WM_DISPLAYCHANGE`?
2. If yes, is there any good way to set the 1st parameter for it, or my way is also OK?
3. If not, could you please let me know the better ways?

Thank you!

**Comment by maxkatz6** [comment, marked as answer]
`WM_DISPLAYCHANGE` is fine, but it's only for Windows.

In current nightly builds, future 11.2 release, we have new `Changed` event on `window.Screens` API, which you can use for this task on most of platforms.

**Reply by Bo-Liu-BL** [reply]
Hi @maxkatz6 ,

Thank you very much for the reply. That's good news for both `WM_DISPLAYCHANGE` is applicable and the new 11.2 support.
In my side, currently we can focus on Windows only.
(And after 11.2 is released, we should be able to switch to it in a short time.)

Given that I call `Win32Properties.AddWndProcHookCallback` in an utility file which doesn't doesn't interact with UI directly, can I use the following style to set the 1st parameter of this method? Or should I use some way to pass the main window handle to the utility layer?

```
if (Avalonia.Application.Current.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime mainWindow)
{
         Win32Properties.AddWndProcHookCallback(mainWindow.MainWindow,
                                                MyWndProcHookCallback);
}
```

Thank you!

**Reply by Bo-Liu-BL** [reply]
Hi @maxkatz6 , is there timeline when we can get 11.2 official release? Thank you!
